03_execution/run-folder.py

```diff
@@ -1,166 +1,211 @@
 # Based on code from https://github.com/standupmaths/xmastree2020
 
+from typing import List, Tuple, Optional
+import time
+import csv
+import os
+import argparse
+
 import board
 import neopixel
-import time 
-from csv import reader
-import sys
-import os
 
-# sleep_time = 0.033 # approx 30fps 
-sleep_time = 0.017  # approx 60fps
-NUMBEROFLEDS = 500
-LOOPS_PER_SEQUENCE = 5
-TRANSITION_FRAMES = 60
-
-if len(sys.argv) > 2:
-    LOOPS_PER_SEQUENCE = int(sys.argv[2])
-if len(sys.argv) > 3:
-    TRANSITION_FRAMES  = int(sys.argv[3])
-
-print("Sequences will loop " + str(LOOPS_PER_SEQUENCE) + " times")
-print("Sequences will blend over " + str(TRANSITION_FRAMES) + " frames")
-
-
-
-
-
-# helper function for chunking 
-def chunks(lst, n):
-    for i in range(0, len(lst), n):
-        yield lst[i:i+n]
-        
-# Given two light frames and a tween ratio, returns an interpolated one
-def tween_frames(frame_a, frame_b, ratio):
-    # sanity
-    len_a = len(frame_a)
-    len_b = len(frame_b)
-    if len_a != len_b:
-        print("cannot interpolate frames with different lengths")
-        return frame_a
-
-    tween = []
-    for i in range(0, len_a):
-        g = round((1 - ratio) * frame_a[i][0] + ratio * frame_b[i][0])
-        r = round((1 - ratio) * frame_a[i][1] + ratio * frame_b[i][1])
-        b = round((1 - ratio) * frame_a[i][2] + ratio * frame_b[i][2])
-        tween.append((g, r, b))  # remember that LEDs take GRB values
-
-    return tween
-
-# Given two sequences, returns the second one with the staring block 
-# blended with the end of the first one
-def blend_lights(lights_a, lights_b, transition_length):
-    # Blend the last and first blocks of lights
-    end_a = lights_a[len(lights_a) - transition_length:]
-    # print(str(len(end_a)))
-    start_b = lights_b[:transition_length]
-    # print(str(len(start_b)))
-
-    blend = []
-    step = 1.0 / (transition_length + 1)
-    for i in range(0, transition_length):
-        n = step * (i + 1)
-        frame = tween_frames(end_a[i], start_b[i], n)
-        blend.append(frame)
-    
-    return blend + lights_b[transition_length:]
-
-
-
-# Given a filename, returns the parsed light object
-def getLights(csvFile):
-    print("Parsing " + csvFile)
-    # read the file
-    # iterate through the entire thing and make all the points the same colour 
-    lightArray = []
-    
-    with open(csvFile, 'r') as read_obj:
+NUMBER_OF_LEDS = 500
+
+
+def parse_animation_csv(
+    csv_path,
+) -> Tuple[List[List[Tuple[float, float, float]]], List[float]]:
```

Comment:
These types are very deeply nested. You could:
- use a `namedtuple` for the outer tuple to make it clearer.
- Alias `Tuple[float, float, float]` as `Color` (maybe use a `namedtuple` for this if it doesn't cause problems elsewhere too)
- Alias `List[Color]` as `Frame`

This would result in the type signature being significantly simpler.

Reply:
My visualiser returns List[List[List[float]], List[float]] so there's not much simplification to be done about this unfortunately but aliases seem like a good way to make this more readable

Reply:
The latest commit resolves this